File: acore_eluna/tree_menu/impl.py
# -*- coding: utf-8 -*-

"""
:class:`LuaCodeGenerator`.
"""

import typing as T
import enum
import dataclasses
import polars as pl


class IconEnum(enum.IntEnum):
    """
    下面是所有 ICON 代码的枚举. 你可以在 "OptionIcon" 一节中看到所有图标的说明.
    See: https://www.azerothcore.org/wiki/gossip_menu_option
    """

    GOSSIP_ICON_CHAT = 0  # White chat bubble
    GOSSIP_ICON_VENDOR = 1  # Brown bag
    GOSSIP_ICON_TAXI = 2  # Flight
    GOSSIP_ICON_TRAINER = 3  # Book
    GOSSIP_ICON_INTERACT_1 = 4  # Interaction wheel
    GOSSIP_ICON_INTERACT_2 = 5  # Interaction wheel
    GOSSIP_ICON_MONEY_BAG = 6  # Brown bag with yellow dot (gold)
    GOSSIP_ICON_TALK = 7  # White chat bubble with black dots (...)
    GOSSIP_ICON_TABARD = 8  # Tabard
    GOSSIP_ICON_BATTLE = 9  # Two swords
    GOSSIP_ICON_DOT = 10  # Yellow dot


class OptionTypeEnum(str, enum.Enum):
    """
    表示一个 Gossip 中可点击的选项.
    """

    ITEM = "item"
    MENU = "menu"
    BACK = "back"


OptionTypeName = T.Literal[
    OptionTypeEnum.ITEM.value,
    OptionTypeEnum.MENU.value,
    OptionTypeEnum.BACK.value,
]


class ItemOptionType(T.TypedDict):
    """
    表示一个 Gossip 中点击了就能产生业务逻辑的选项.
    """

    id: int
    name: str
    type: OptionTypeName
    icon: int
    parent: int
    data: T.Dict[str, T.Any]


class MenuOptionType(T.TypedDict):
    """
    表示一个 Gossip 中点击了就进入下级菜单的选项.
    """

    id: int
    name: str
    type: OptionTypeName
    icon: int
    parent: int


class BackOptionType(T.TypedDict):
    """
    表示一个 Gossip 中点击了就进入回到上一级或者主菜单的选项.
    """

    id: int
    name: str
    type: OptionTypeName
    icon: int
    parent: int
    back_to: int


OptionType = T.Union[ItemOptionType, MenuOptionType]


ROOT_PARENT_ID = 0


@dataclasses.dataclass
class LuaCodeGenerator:
    """
    这个类能帮助你将一个在 Google sheet 中定义的 dataframe 转化成 Lua 中所需要的
    gossip menu 的数据.

    :param id_start: The starting id for the menu item. If the value is 1,
        then your menu id will be 1, 2, 3, .... If you have multiple gossip
        menus, you have to ensure their id does not overlap.
    :param item_option_icon: The icon for the item. If the ``_icon`` field in the
        data frame is None, then this icon will be used.
    :param menu_option_icon: The icon for the menu. If the ``_icon`` field in the
        data frame is None, then this icon will be used.
    """

    id_start: int = dataclasses.field()
    item_option_icon: IconEnum = dataclasses.field(default=IconEnum.GOSSIP_ICON_CHAT)
    menu_option_icon: IconEnum = dataclasses.field(default=IconEnum.GOSSIP_ICON_CHAT)
    back_option_icon: IconEnum = dataclasses.field(default=IconEnum.GOSSIP_ICON_TALK)
    back_to_prev: str = dataclasses.field(default="Back to {parent_name}")
    back_to_top: str = dataclasses.field(default="Back to Top")

    def __post_init__(self):
        if self.id_start < 1:
            raise ValueError("The id_start must be greater than 0.")

    def make_item_option(
        self,
        row: T.Dict[str, T.Any],
        header: int,
        parent: int = ROOT_PARENT_ID,
    ) -> ItemOptionType:
        """
        一个 item option 就是一个 Gossip 中点击了就能产生业务逻辑的选项.
        """
        if row["_icon"] is None:
            icon = self.item_option_icon.value
        else:  # pragma: no cover
            icon = row["_icon"]
        item_option = {
            "id": self.id_start,
            "name": row[f"_h{header}"],
            "type": OptionTypeEnum.ITEM.value,
            "icon": icon,
            "parent": parent,
        }
        self.id_start += 1
        for k in list(row):
            if k.startswith("_"):
                del row[k]
        item_option["data"] = row
        return item_option

    def make_menu_option(
        self,
        name: str,
        parent: int = ROOT_PARENT_ID,
    ) -> MenuOptionType:
        """
        一个 menu option 就是一个 Gossip 中点击了就进入下级菜单的选项.
        """
        menu_option = {
            "id": self.id_start,
            "name": name,
            "type": OptionTypeEnum.MENU.value,
            "icon": self.menu_option_icon.value,
            "parent": parent,
        }
        self.id_start += 1
        return menu_option

    def make_back_option(
        self,
        name: str,
        parent: int = ROOT_PARENT_ID,
        back_to: int = ROOT_PARENT_ID,
    ) -> BackOptionType:
        """
        一个 back option 就是一个 Gossip 中点击了就进入回到上一级或者主菜单的选项.
        """
        back_option = {
            "id": self.id_start,
            "name": name,
            "type": OptionTypeEnum.BACK.value,
            "icon": self.back_option_icon.value,
            "parent": parent,
            "back_to": back_to,
        }
        self.id_start += 1
        return back_option

    def _dataframe_to_menu_data(
        self,
        df: pl.DataFrame,
        _parent: T.Optional[int] = None,
        _parent_name: T.Optional[str] = None,
        _header: int = 1,
        _items: T.Optional[T.List[T.Dict[str, T.Any]]] = None,
    ) -> T.List[OptionType]:
        """
        输入一个 dataframe, 自动根据层级关系生成 gossip menu 中的 option,
        返回一个列表, 里面的每个元素都是一个按钮.

        这个函数用到了递归实现.

        :param df: see sample table
            https://docs.google.com/spreadsheets/d/1ZLNDemVn_5T1GbZZPd2igMCkPmEpZ1tR5U28nEkE41Y/edit?gid=0#gid=0
        """
        if _items is None:
            _items = list()
        if f"_h{_header + 1}" not in df.schema:
            for row in df.to_dicts():
                item = self.make_item_option(row, header=_header, parent=_parent)
                _items.append(item)
            if _parent is not None:
                _items.extend(
                    [
                        self.make_back_option(
                            name=self.back_to_prev.format(parent_name=_parent_name),
                            parent=_parent,
                            back_to=_parent,
                        ),
                        self.make_back_option(
                            name=self.back_to_top,
                            parent=_parent,
                            back_to=ROOT_PARENT_ID,
                        ),
                    ]
                )
            return _items

        # 如果下级菜单 (下个 header) 没有数据, 那么这是一个 menu, 如果有数据, 那么这是一个 item
        df_menu = df.filter(pl.col(f"_h{_header + 1}").is_not_null())
        df_vendor = df.filter(pl.col(f"_h{_header + 1}").is_null())

        # 我们优先创建 menu
        for (name,), sub_df in df_menu.group_by([f"_h{_header}"], maintain_order=True):
            menu = self.make_menu_option(name=name, parent=_parent)
            _items.append(menu)
            self._dataframe_to_menu_data(
                df=sub_df,
                _parent=menu["id"],
                _parent_name=menu["name"],
                _header=_header + 1,
                _items=_items,
            )

        # 然后创建 item
        for row in df_vendor.to_dicts():
            item = self.make_item_option(row, header=_header, parent=_parent)
            _items.append(item)

        if _parent is not None:
            _items.extend(
                [
                    self.make_back_option(
                        name=self.back_to_prev.format(parent_name=_parent_name),
                        parent=_parent,
                        back_to=_parent,
                    ),
                    self.make_back_option(
                        name="Back to Top",
                        parent=_parent,
                        back_to=ROOT_PARENT_ID,
                    ),
                ]
            )

        return _items

    def dataframe_to_option_list(
        self,
        df: pl.DataFrame,
        _parent: int = ROOT_PARENT_ID,
        _header: int = 1,
        _items: T.Optional[T.List[T.Dict[str, T.Any]]] = None,
    ) -> T.List[OptionType]:
        """
        输入一个 dataframe, 自动根据层级关系生成 gossip menu 中的 option 列表.
        里面的每个元素都是一个按钮.

        这个函数用到了递归实现.

        :param df: see sample table
            https://docs.google.com/spreadsheets/d/1ZLNDemVn_5T1GbZZPd2igMCkPmEpZ1tR5U28nEkE41Y/edit?gid=0#gid=0
        """
        # validate input dataframe
        url = "https://docs.google.com/spreadsheets/d/1ZLNDemVn_5T1GbZZPd2igMCkPmEpZ1tR5U28nEkE41Y/edit?gid=0#gid=0"
        msg = f" See {url} for a good example."
        if "_icon" not in df.schema:  # pragma: no cover
            raise ValueError(
                f"The input dataframe must have a column named '_icon'.{msg}"
            )
        if "_h1" not in df.schema:  # pragma: no cover
            raise ValueError(
                f"The input dataframe must have a column named '_h1'.{msg}"
            )
        h_columns = [int(k[2:]) for k in df.schema if k.startswith("_h")]
        h_columns.sort()
        if all(df[f"_h{h_columns[-1]}"].is_null()):  # pragma: no cover
            raise ValueError(f"The last header column must NOT be ALL NULL.{msg}")
        columns = set(df.schema)
        columns.remove("_icon")
        columns = columns.difference([f"_h{col}" for col in h_columns])
        for col in columns:
            if col.startswith("_"):  # pragma: no cover
                raise ValueError(f"The data column name must NOT start with '_'.{msg}")

        return self._dataframe_to_menu_data(
            df=df,
            _parent=_parent,
            _header=_header,
            _items=_items,
        )

    def generate_lua_code(
        self,
        option_list: T.List[OptionType],
        data_to_lua_code: T.Callable[[T.Dict[str, T.Any]], str],
    ) -> str:
        """
        根据 option list 生成业务数据代码.

        例如如果 ``option_list`` 长这个样子:

        .. code-block:: python

            [
                {'id': 1001, 'name': '牧师', 'type': 'menu', 'icon': 3, 'parent': 0},
                {
                    'id': 1002,
                    'name': '真言术韧',
                    'type': 'menu',
                    'icon': 3,
                    'parent': 1001
                },
                {
                    'id': 1003,
                    'name': '真言术韧 60',
                    'type': 'item',
                    'icon': 9,
                    'parent': 1002,
                    'data': {'buff_id': 10938, 'buff_count': 1}
                },
                {
                    'id': 1004,
                    'name': '真言术韧 70',
                    'type': 'item',
                    'icon': 9,
                    'parent': 1002,
                    'data': {'buff_id': 25389, 'buff_count': 1}
                },
                {
                    'id': 1005,
                    'name': '真言术韧 80',
                    'type': 'item',
                    'icon': 9,
                    'parent': 1002,
                    'data': {'buff_id': 48161, 'buff_count': 1}
                },
                {
                    'id': 1006,
                    'name': 'Back to 真言术韧',
                    'type': 'back',
                    'icon': 7,
                    'parent': 1002,
                    'back_to': 1002
                },
                {
                    'id': 1007,
                    'name': 'Back to Top',
                    'type': 'back',
                    'icon': 7,
                    'parent': 1002,
                    'back_to': 0
                },
                {
                    'id': 1008,
                    'name': '反恐惧结界',
                    'type': 'item',
                    'icon': 9,
                    'parent': 1001,
                    'data': {'buff_id': 6346, 'buff_count': 1}
                },
                {
                    'id': 1009,
                    'name': '能量灌注',
                    'type': 'item',
                    'icon': 9,
                    'parent': 1001,
                    'data': {'buff_id': 10060, 'buff_count': 1}
                },
                {
                    'id': 1010,
                    'name': 'Back to 牧师',
                    'type': 'back',
                    'icon': 7,
                    'parent': 1001,
                    'back_to': 1001
                },
                {
                    'id': 1011,
                    'name': 'Back to Top',
                    'type': 'back',
                    'icon': 7,
                    'parent': 1001,
                    'back_to': 0
                },
                {
                    'id': 1012,
                    'name': '王者祝福',
                    'type': 'item',
                    'icon': 9,
                    'parent': 0,
                    'data': {'buff_id': 56525, 'buff_count': 1}
                },
                {
                    'id': 1013,
                    'name': 'Back to None',
                    'type': 'back',
                    'icon': 7,
                    'parent': 0,
                    'back_to': 0
                },
                {
                    'id': 1014,
                    'name': 'Back to Top',
                    'type': 'back',
                    'icon': 7,
                    'parent': 0,
                    'back_to': 0
                }
            ]

        那么生成的 lua 代码就是这个样子:

        .. code-block:: lua

            { id = 1001, name = "牧师", type = "menu", icon = 3, parent = 0 },
            { id = 1002, name = "真言术韧", type = "menu", icon = 3, parent = 1001 },
            { id = 1003, name = "真言术韧 60", type = "item", icon = 9, parent = 1002, data = { buff_id = 10938, buff_count = 1 } },
            { id = 1004, name = "真言术韧 70", type = "item", icon = 9, parent = 1002, data = { buff_id = 25389, buff_count = 1 } },
            { id = 1005, name = "真言术韧 80", type = "item", icon = 9, parent = 1002, data = { buff_id = 48161, buff_count = 1 } },
            { id = 1006, name = "Back to 真言术韧", type = "back", icon = 7, parent = 1002, back_to = 1002 },
            { id = 1007, name = "Back to Top", type = "back", icon = 7, parent = 1002, back_to = 0 },
            { id = 1008, name = "反恐惧结界", type = "item", icon = 9, parent = 1001, data = { buff_id = 6346, buff_count = 1 } },
            { id = 1009, name = "能量灌注", type = "item", icon = 9, parent = 1001, data = { buff_id = 10060, buff_count = 1 } },
            { id = 1010, name = "Back to 牧师", type = "back", icon = 7, parent = 1001, back_to = 1001 },
            { id = 1011, name = "Back to Top", type = "back", icon = 7, parent = 1001, back_to = 0 },
            { id = 1012, name = "王者祝福", type = "item", icon = 9, parent = 0, data = { buff_id = 56525, buff_count = 1 } },
            { id = 1013, name = "Back to None", type = "back", icon = 7, parent = 0, back_to = 0 },
            { id = 1014, name = "Back to Top", type = "back", icon = 7, parent = 0, back_to = 0 },

        ``data_to_lua_code`` 参数是一个函数, 用来将你的核心业务数据转化成 lua 的源代码.
        也就是上面例子中把 ``{"buff_id": 42995, "buff_count": 1}`` 转化成
        ``{ buff_id = 42995, buff_count = 1 }``.
        因为我们无法预测用户的核心业务数据长什么样, 所以只能交给用户来自己实现.
        """
        lines = list()
        tab = " " * 4
        for option in option_list:
            id = option["id"]
            name = option["name"]
            type = option["type"]
            icon = option["icon"]
            parent = option["parent"]
            if type == OptionTypeEnum.ITEM.value:
                data = data_to_lua_code(option["data"])
                line = (
                    f"{tab}{{ "
                    f"id = {id}, "
                    f'name = "{name}", '
                    f'type = "{type}", '
                    f"icon = {icon}, "
                    f"parent = {parent}, "
                    f"data = {data} "
                    f"}},"
                )
            elif type == OptionTypeEnum.MENU.value:
                line = (
                    f"{tab}{{ "
                    f"id = {id}, "
                    f'name = "{name}", '
                    f'type = "{type}", '
                    f"icon = {icon}, "
                    f"parent = {parent} "
                    f"}},"
                )
            elif type == OptionTypeEnum.BACK.value:
                back_to = option["back_to"]
                line = (
                    f"{tab}{{ "
                    f"id = {id}, "
                    f'name = "{name}", '
                    f'type = "{type}", '
                    f"icon = {icon}, "
                    f"parent = {parent}, "
                    f"back_to = {back_to} "
                    f"}},"
                )
            else:  # pragma: no cover
                raise TypeError(f"Unknown type {type!r}")
            lines.append(line)
        return "\n".join(lines)

File: acore_eluna/tree_menu/impl.lua
--[[
这个例子演示了如果用纯 Lua 实现的话, 代码应该是怎样的.

首先这里有几个难点:

1. MENU_DATA_TREE 这个层级结构的 Table 不利于人类输入.
2. 我们需要 Preprocess 这个递归函数将 MENU_DATA_TREE 转换为平铺的 MENU_DATA_MAPPER.

这也是我们为什么要创造 tree_menu Python 库的原因. 它的好处是:

1. 不再需要 MENU_DATA_TREE, 我们直接在 Excel 表格中进行编辑.
2. 不再需要 Preprocess 函数, Python 代码会自动将数据转换为平铺的 MENU_DATA_MAPPER.
3. Python 会自动生成 Lua 中的 MENU_DATA_MAPPER 代码, 你只需要复制粘贴既可.
--]]

local ROOT_PARENT_ID = 0
local GOSSIP_ICON_TRAINER = 3 -- Book
local GOSSIP_ICON_BATTLE = 9 -- Two swords

--[[
MENU_DATA_TREE

把你希望给玩家看到的传送菜单的数据按照层级结构放在这个列表中. 这里的每条记录叫做一个 menuData.
一条 menuData 对应着传送菜单上的一个按钮, 也对应着一个传送坐标.
--]]
local MENU_DATA_TREE = {
    {
        { name = "牧师" },
        {
            { name = "真言术: 韧" },
            { name = "真言术韧 60", buff_id = 10938, buff_count = 1 },
            { name = "真言术韧 70", buff_id = 25389, buff_count = 1 },
            { name = "真言术韧 80", buff_id = 48161, buff_count = 1 },
        },
    },
    {
        { name = "法师" },
        {
            { name = "奥术智慧" },
            { name = "奥术智慧 60", buff_id = 10157, buff_count = 1 },
            { name = "奥术智慧 70", buff_id = 27126, buff_count = 1 },
            { name = "奥术智慧 80", buff_id = 42995, buff_count = 1 },
        },
    },
    {
        { name = "战斗" },
        { name = "兽群领袖光环", buff_id = 24932, buff_count = 1 },
        { name = "枭兽光环", buff_id = 24907, buff_count = 1 },
    },
    { name = "王者祝福", buff_id = 56525, buff_count = 1 },
}

local MENU_DATA_MAPPER = {}
local current_id = 1
local function Preprocess(menuDataTree, parentMenuDataId)
    --[[
    这个函数是对 MENU_DATA_TREE 表中的数据进行解析, 把层级树结构转换为平铺的 key, value
    mapper 结构 (MENU_DATA_MAPPER 这个变量). 这个函数用到了递归.

    :param menuDataTree: 这是一个列表, 里面的元素是类似于
        {name = "达拉然 飞行管理员", map = 571, x = 5813.0, y = 448.0, z = 658.8, o = 0}
        这样的字典.
    :param parentMenuDataId: 这是一个数字, 用于标识当前 menuDataList 的父级菜单.
    --]]
    -- 类似于 Python 中的 enumerate 函数, 返回一个索引和值的元组
    for ind, menuData in ipairs(menuDataTree) do
        newMenuData = {
            id = current_id,
            name = menuData.name,
            is_menu = false,
            icon = GOSSIP_ICON_BATTLE,
            parent_id = parentMenuDataId,
            data = menuData,
        }
        current_id = current_id + 1
        MENU_DATA_MAPPER[newMenuData.id] = newMenuData
        --print(string.format("newMenuData.id = %s, newMenuData.name = %s, newMenuData.is_menu = %s, newMenuData.parent_id = %s", newMenuData.id, newMenuData.name, newMenuData.is_menu, newMenuData.parent_id))

        -- 如果 menuData 是一个菜单, 那么递归调用这个函数
        if not menuData.buff_id then
            MENU_DATA_MAPPER[newMenuData.id].is_menu = true
            MENU_DATA_MAPPER[newMenuData.id].icon = GOSSIP_ICON_TRAINER
            Preprocess(menuData, newMenuData.id)
        end
    end
end

Preprocess(MENU_DATA_TREE, ROOT_PARENT_ID)

local function FindIdByKeyValue(menuDataMapper, menuDataKey, menuDataValue)
    --[[
    这个函数的目的是查找第一个 key, value pair 符合条件的 menuData 的 ID.

    类似于 ``SELECT ID FROM table WHERE talbe.menuDataKey = menuDataValue LIMIT 1``.

    :type menuDataMapper: table
    :param menuDataMapper: 一个 {ID: menuData} 的字典, 其中 ID 是整数.
    :type menuDataKey: string
    :param menuDataKey: menuData 中的 key
    :type menuDataKey: any
    :param menuDataValue: menuData 中的 value

    :return: 符合条件的 menuData 的 ID.
    --]]
    for menuDataId, menuData in pairs(menuDataMapper) do
        if menuDataKey then
            if menuData[menuDataKey] == menuDataValue then
                return menuDataId
            end
        else
            -- 貌似无论如何都不会进入到这段逻辑中
            if menuData == menuDataValue then
                return menuDataId
            end
        end
    end
end

local function FindAllByKeyValue(menuDataMapper, menuDataKey, menuDataValue)
    --[[
    这个函数的目的是查找所有 key, value pair 符合条件的 menuData (不是 ID 而是数据) 的列表.
    类似于 ``SELECT * FROM table WHERE talbe.menuDataKey = menuDataValue``.
    这个函数主要用于根据 parent_id 查找所有子菜单.

    :type menuDataMapper: table
    :param menuDataMapper: 一个 {ID: menuData} 的字典, 其中 ID 是整数.
    :type menuDataKey: string
    :param menuDataKey: menuData 中的 key
    :type menuDataKey: any
    :param menuDataValue: menuData 中的 value

    :return: 符合条件的所有 menuData 的列表.
    --]]
    local menuDataList = {}
    for menuDataId, menuData in pairs(menuDataMapper) do
        if menuDataKey then
            if menuData[menuDataKey] == menuDataValue then
                table.insert(menuDataList, menuData)
            end
        else
            if menuData == menuDataValue then
                table.insert(menuDataList, menuData)
            end
        end
    end
    return menuDataList
end

local function PrintMenuData(menuData)
    print(string.format("---------- %s ----------", menuData.id))
    print(string.format("id = %s", menuData.id))
    print(string.format("name = %s", menuData.name))
    print(string.format("is_menu = %s", menuData.is_menu))
    print(string.format("icon = %s", menuData.icon))
    print(string.format("parent_id = %s", menuData.parent_id))
    print(string.format("data.buff_id = %s", menuData.data.buff_id))
    print(string.format("data.buff_count = %s", menuData.data.buff_count))
end

print("========== MENU_DATA_MAPPER ==========")
for id, menuData in pairs(MENU_DATA_MAPPER) do
    PrintMenuData(menuData)
end

print("========== FindIdByKeyValue ==========")
local menuDataId = FindIdByKeyValue(MENU_DATA_MAPPER, "name", "真言术韧 80")
PrintMenuData(MENU_DATA_MAPPER[menuDataId])

print("========== FindAllByKeyValue ==========")
local menuDataList = FindAllByKeyValue(MENU_DATA_MAPPER, "parent_id", 11)
for _, menuData in pairs(menuDataList) do
    PrintMenuData(menuData)
end
